# Create project

## Import Libraries

In [17]:
# import libraries
## Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
## Importing the OS and JSON Modules
import os,json

## Load data

In [84]:
# make the dataframe
superhero= pd.read_csv('superhero_info - superhero_info.csv')

In [85]:
# check the data loaded
superhero.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [86]:
# look at the basicinfo
superhero.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


In [104]:
#load the second dataframe
superpower=pd.read_csv('superhero_powers - superhero_powers.csv')

In [105]:
#check that the data loaded
superpower.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [89]:
# check the info
superpower.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


## Clean and prepare the data

In [90]:
#check for duplicates
superhero.duplicated().sum()

0

In [91]:
#check for duplicates
superpower.duplicated().sum()

0

- There are 0 duplicates. 

### Divide Hero/Publisher column

In [92]:
# Exploring existing format with a few examples
superhero['Hero|Publisher'].head(2)

0            A-Bomb|Marvel Comics
1    Abe Sapien|Dark Horse Comics
Name: Hero|Publisher, dtype: object

In [93]:
## examining a single value from the coordinates col
info = superhero.loc[0,"Hero|Publisher"]
print(type(info))
info

<class 'str'>


'A-Bomb|Marvel Comics'

In [94]:
## adding expand=True
superhero['Hero|Publisher'].str.split('|',expand=True)

0                  1
0             A-Bomb      Marvel Comics
1         Abe Sapien  Dark Horse Comics
2           Abin Sur          DC Comics
3        Abomination      Marvel Comics
4      Absorbing Man      Marvel Comics
..               ...                ...
458     Yellowjacket      Marvel Comics
459  Yellowjacket II      Marvel Comics
460             Yoda       George Lucas
461          Zatanna          DC Comics
462             Zoom          DC Comics

[463 rows x 2 columns]

In [100]:
## save the 2 new columns into the dataframe
superhero[['Hero','Publisher']] = superhero['Hero|Publisher'].str.split('|',expand=True)
superhero.head(2)

Hero|Publisher Gender           Race Alignment Hair color  \
0          A-Bomb|Marvel Comics   Male          Human      good    No Hair   
1  Abe Sapien|Dark Horse Comics   Male  Icthyo Sapien      good    No Hair   

  Eye color Skin color                                  Measurements  \
0    yellow    Unknown  {'Height': '203.0 cm', 'Weight': '441.0 kg'}   
1      blue       blue   {'Height': '191.0 cm', 'Weight': '65.0 kg'}   

         Hero          Publisher  
0      A-Bomb      Marvel Comics  
1  Abe Sapien  Dark Horse Comics

In [101]:
# drop the Hero|Publisher 
superhero = superhero.drop(columns=['Hero|Publisher'])
superhero.head(2)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements        Hero          Publisher  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb      Marvel Comics  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien  Dark Horse Comics

### Divide the measurements column 

In [18]:
# Exploring existing format with a few examples
superhero['Measurements'].head(2)

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
Name: Measurements, dtype: object

In [107]:
## examining a single value from the measurement col
measure = superhero.loc[0,"Measurements"]
print(type(measure))
measure

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [108]:
measure = measure.replace("'",'"')
measure

'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [109]:
## now we can use json.loads
fixed_measure = json.loads(measure)
print(type(fixed_measure))
fixed_measure

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [110]:
## use .str.replace to replace all single quotes
superhero['Measurements'] = superhero['Measurements'].str.replace("'",'"')
## Apply the json.loads to the full column
superhero['Measurements'] = superhero['Measurements'].apply(json.loads)
superhero['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [111]:
#create a column from the dictionary
height_weight = superhero['Measurements'].apply(pd.Series)
height_weight

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [113]:
# concat height_weight with original dataframe
superhero = pd.concat((superhero, height_weight), axis = 1)
superhero.head(2)


Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements        Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien   

           Publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg

In [167]:
#fix the weight column
# Make a list of all characters to replace
to_replace = [' kg',]
# run a loop to replace all of the characters in the list at once
for char in to_replace:
    superhero['Weight'] = superhero['Weight'].str.replace(char,'',regex=False)
    
superhero['Weight'].head(5)

0    441.0
1     65.0
2     90.0
3    441.0
4    122.0
Name: Weight, dtype: object

In [168]:
# convert to a float
superhero['Weight'] = superhero['Weight'].astype(float)

In [169]:
#fix the height column
# Make a list of all characters to replace
to_replace = [' cm',]
# run a loop to replace all of the characters in the list at once
for char in to_replace:
    superhero['Height'] = superhero['Height'].str.replace(char,'',regex=False)
    
superhero['Height'].head(5)

0    203.0
1    191.0
2    185.0
3    203.0
4    193.0
Name: Height, dtype: object

In [170]:
# convert to a float
superhero['Height'] = superhero['Height'].astype(float)

In [ ]:
# drop the original column it was already dropped
superhero = superhero.drop(columns=['Measurements'])

### Encode the superpower

In [116]:
# Exploring existing format with a few examples
superpower['Powers'].head(2)

0           Agility,Super Strength,Stamina,Super Speed
1    Accelerated Healing,Durability,Longevity,Super...
Name: Powers, dtype: object

In [123]:
## showing the lists are really strings
superpower.loc[2,'Powers']

'Agility,Accelerated Healing,Cold Resistance,Durability,Underwater breathing,Marksmanship,Weapons Master,Longevity,Intelligence,Super Strength,Telepathy,Stamina,Immortality,Reflexes,Enhanced Sight,Sub-Mariner'

In [138]:
## slice out a test power
test_power = superpower.loc[5, 'Powers']
test_power

'Dimensional Awareness,Flight,Intelligence,Super Strength,Size Changing,Super Speed,Teleportation,Magic,Dimensional Travel,Immortality,Invulnerability,Molecular Manipulation,Energy Manipulation,Power Cosmic'

In [139]:
#check the type
type(test_power)


str

In [140]:
# Replacing single ' with double "
saved_test = test_power.replace("'", '"')
saved_test

'Dimensional Awareness,Flight,Intelligence,Super Strength,Size Changing,Super Speed,Teleportation,Magic,Dimensional Travel,Immortality,Invulnerability,Molecular Manipulation,Energy Manipulation,Power Cosmic'

In [146]:
# Create a new column where the single quotes are replaced by double quotes
superpower['powers_split'] = superpower['Powers'].str.split(',')

In [154]:
## exploding the column of lists
exploded = superpower.explode('powers_split')
exploded.head(20)

hero_names                                             Powers  \
0     3-D Man         Agility,Super Strength,Stamina,Super Speed   
0     3-D Man         Agility,Super Strength,Stamina,Super Speed   
0     3-D Man         Agility,Super Strength,Stamina,Super Speed   
0     3-D Man         Agility,Super Strength,Stamina,Super Speed   
1      A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1      A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1      A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1      A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1      A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1      A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1      A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   

           powers_split  Agility  Super Strength  Stamina  Super Speed  \
0               Agility     True            True     True         True   
0        Super Strength     True            True     True         True   
0               Stamina     True            True     True         True   
0           Super Speed     True            True     True         True   
1   Accelerated Healing    False            True     True        False   
1            Durability    False            True     True        False   
1             Longevity    False            True     True        False   
1        Super Strength    False            True     True        False   
1               Stamina    False            True     True        False   
1            Camouflage    False            True     True        False   
1       Self-Sustenance    False            True     True        False   
2               Agility     True            True     True        False   
2   Accelerated Healing     True            True     True        False   
2       Cold Resistance     True            True     True        False   
2            Durability     True            True     True        False   
2  Underwater breathing     True            True     True        False   
2          Marksmanship     True            True     True        False   
2        Weapons Master     True            True     True        False   
2             Longevity     True            True     True        False   
2          Intelligence     True            True     True        False   

   Accelerated Healing  Durability  Longevity  ...  Weather Control  \
0                False       False      False  ...            False   
0                False       False      False  ...            False   
0                False       False      False  ...            False   
0                False       False      False  ...            False   
1                 True        True       True  ...            False   
1                 True        True       True  ...            False   
1                 True        True       True  ...            False   
1                 True        True       True  ...            False   
1                 True        True       True  ...            False   
1                 True        True       True  ...            False   
1                 True        True       True  ...            False   
2                 True        True       True  ...            False   
2                 True        True       True  ...            False   
2                 True   

In [155]:
## saving the unique values from the exploded column
cols_to_make = exploded['powers_split'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [156]:
superpower.head(20)

hero_names                                             Powers  \
0            3-D Man         Agility,Super Strength,Stamina,Super Speed   
1             A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2         Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3           Abin Sur                                 Lantern Power Ring   
4        Abomination  Accelerated Healing,Intelligence,Super Strengt...   
5            Abraxas  Dimensional Awareness,Flight,Intelligence,Supe...   
6      Absorbing Man  Cold Resistance,Durability,Energy Absorption,S...   
7        Adam Monroe       Accelerated Healing,Immortality,Regeneration   
8       Adam Strange  Durability,Stealth,Flight,Marksmanship,Weapons...   
9          Agent Bob                                            Stealth   
10        Agent Zero  Energy Absorption,Longevity,Super Strength,Ene...   
11        Air-Walker  Flight,Super Strength,Super Speed,Self-Sustena...   
12              Ajax  Agility,Super Strength,Super Speed,Heat Genera...   
13        Alan Scott  Flight,Danger Sense,Intelligence,Telepathy,Ene...   
14       Alex Mercer  Agility,Accelerated Healing,Stealth,Energy Abs...   
15      Alex Woolsly                               Underwater breathing   
16             Alien  Agility,Cold Resistance,Durability,Stealth,Sup...   
17  Allan Quatermain                                       Marksmanship   
18             Amazo  Agility,Lantern Power Ring,Durability,Flight,U...   
19              Ammo                                     Weapons Master   

                                         powers_split  Agility  \
0     [Agility, Super Strength, Stamina, Super Speed]     True   
1   [Accelerated Healing, Durability, Longevity, S...    False   
2   [Agility, Accelerated Healing, Cold Resistance...     True   
3                                [Lantern Power Ring]    False   
4   [Accelerated Healing, Intelligence, Super Stre...    False   
5   [Dimensional Awareness, Flight, Intelligence, ...    False   
6   [Cold Resistance, Durability, Energy Absorptio...    False   
7    [Accelerated Healing, Immortality, Regeneration]    False   
8   [Durability, Stealth, Flight, Marksmanship, We...    False   
9                                           [Stealth]    False   
10  [Energy Absorption, Longevity, Super Strength,...    False   
11  [Flight, Super Strength, Super Speed, Self-Sus...    False   
12  [Agility, Super Strength, Super Speed, Heat Ge...     True   
13  [Flight, Danger Sense, Intelligence, Telepathy...    False   
14  [Agility, Accelerated Healing, Stealth, Energy...     True   
15                             [Underwater breathing]    False   
16  [Agility, Cold Resistance, Durability, Stealth...     True   
17                                     [Marksmanship]    False   
18  [Agility, Lantern Power Ring, Durability, Flig...     True   
19                                   [Weapons Master]    False   

    Super Strength  Stamina  Super Speed  Accelerated Healing  Durability  \
0             True     True         True                False       False   
1             True     True        False                 True        True   
2             True     True        False                 True        True   
3            False    False        False                False       False   
4             True     True         True                 True       False   
5             True    False         True                False       False   
6             True    False        False                False        True   
7            False    False        False                 True       False   
8            False    False         True                False        True   
9            False    False        False                False       False   
10            True    False        False                False       False   
11            True    False         True                False       False   
12            True    False         True      

In [157]:
#create new columns
for col in cols_to_make:
    superpower[col] = superpower['Powers'].str.contains(col)
superpower.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        powers_split  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

In [159]:
# drop powers columns
superpower = superpower.drop(columns=['Powers','powers_split'])
## save data for next step
superpower.to_csv('superpower_all.csv', index=False)

In [162]:
superpower.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Columns: 168 entries, hero_names to Changing Armor
dtypes: bool(167), object(1)
memory usage: 114.1+ KB


### Combine the dataframes

In [177]:
#read the new superpower
superpower_all=pd.read_csv('superpower_all.csv')

In [178]:
## example merge with our 2 dataframes
super_merged = pd.merge(superhero, superpower_all, left_on='Hero', right_on='hero_names', how='inner')
super_merged.head(3)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   
2   Male        Ungaran      good    No Hair      blue        red    Abin Sur   

           Publisher  Height  Weight  ... Weather Control  Omnipresent  \
0      Marvel Comics   203.0   441.0  ...           False        False   
1  Dark Horse Comics   191.0    65.0  ...           False        False   
2          DC Comics   185.0    90.0  ...           False        False   

   Omniscient  Hair Manipulation  Nova Force  Odin Force  Phoenix Force  \
0       False              False       False       False          False   
1       False              False       False       False          False   
2       False              False       False       False          False   

   Intuitive aptitude  Melting  Changing Armor  
0               False    False           False  
1               False    False           False  
2               False    False           False  

[3 rows x 178 columns]

# Answer the questions

- Compare the average weight of super powers who have Super Speed to those who do not.
- What is the average height of heroes for each publisher?

In [179]:
# check column names
super_merged['Super Speed'].value_counts()

False    265
True     198
Name: Super Speed, dtype: int64

In [180]:
#  use groupby to  find average weight of superspeed versus those who do not
super_merged.groupby(['Super Speed'])['Weight'].mean()

Super Speed
False    101.773585
True     129.404040
Name: Weight, dtype: float64

- The average Weight of non-superspeed is 101.77 kg. 
- The average weight of superspeed is 129.40 kg. 

In [181]:
# using groupby
super_merged.groupby(['Publisher'])['Height'].mean()

Publisher
DC Comics            181.923913
Dark Horse Comics    176.909091
George Lucas         159.600000
Image Comics         211.000000
Marvel Comics        191.546128
Shueisha             171.500000
Star Trek            181.500000
Team Epic TV         180.750000
Unknown              178.000000
Name: Height, dtype: float64